NetKet/Hilbert/abstract_hilbert.hpp

```diff
@@ -80,7 +105,10 @@ class AbstractHilbert {
 
   virtual const AbstractGraph &GetGraph() const noexcept = 0;
 
-  virtual ~AbstractHilbert() {}
+  virtual ~AbstractHilbert() = default;
+
+ private:
+  mutable nonstd::optional<HilbertIndex> index_;
```

Comment:
mutable? why?

Reply:
Calling `GetIndex()` should also work on a const reference (creating the HilbertIndex does not change any observable property of the Hilbert object). However, since `index_` needs to be modified the first time `GetIndex` is called, this would not work on a const object, unless it is declared as mutable.

The other option would be to create the HilbertIndex in the constructor, but then it will also be created even when it is never used (which could be ok, creating it on demand like here is just an optimization).

Reply:
See also here: https://isocpp.org/wiki/faq/const-correctness#mutable-data-members

Reply:
I see cool, then there is many other places we should use that too :)